3.optimizing_CP/run_sam/scripts/run_sam2_microscopy.py

```diff
@@ -0,0 +1,673 @@
+        plt.show()
+
+    # must return at least 1 valid coord
+    if valid_coords.shape[0] == 0:
+        valid_coords, valid_labels = generate_random_coords(
+            img=img, coords=coords, samples=samples
+        )
+
+    return valid_coords, valid_labels
+
+
+# ### Memory profiling for GPU
+
+# In[4]:
+
+
+def start_record_memory_history() -> None:
+    """
+    Start recording memory history.
+    """
+    if not torch.cuda.is_available():
+        logger.info("CUDA unavailable. Not recording memory history")
+        return
+
+    logger.info("Starting snapshot record_memory_history")
+    torch.cuda.memory._record_memory_history(
+        max_entries=MAX_NUM_OF_MEM_EVENTS_PER_SNAPSHOT
+    )
+
+
+def stop_record_memory_history() -> None:
+    """
+    Stop recording memory history.
+    """
+    if not torch.cuda.is_available():
+        logger.info("CUDA unavailable. Not recording memory history")
+        return
+
+    logger.info("Stopping snapshot record_memory_history")
+    torch.cuda.memory._record_memory_history(enabled=None)
+
+
+def export_memory_snapshot(save_dir: pathlib.Path) -> None:
+    """
+    Export memory snapshot.
+    """
+    if not torch.cuda.is_available():
+        logger.info("CUDA unavailable. Not exporting memory snapshot")
+        return
+
+    # Prefix for file names.
+    host_name = socket.gethostname()
+    timestamp = datetime.now().strftime(TIME_FORMAT_STR)
+    file_prefix = f"{host_name}_{timestamp}"
+    full_file_path = pathlib.Path(save_dir / f"{file_prefix}.pickle").resolve()
+    save_dir.mkdir(parents=True, exist_ok=True)
+
+    try:
+        logger.info(f"Saving snapshot to local file: {full_file_path}")
+        torch.cuda.memory._dump_snapshot(full_file_path)
+    except Exception as e:
+        logger.error(f"Failed to capture memory snapshot {e}")
+        return
+
+
+def delete_recorded_memory_history(save_dir: pathlib.Path) -> None:
+    """
+    Delete recorded memory history.
+    """
+    file_path = pathlib.Path(save_dir).resolve()
+    files = list(file_path.glob("*.pickle"))
+    try:
+        [file.unlink() for file in files]
+        logger.info(f"Deleted memory snapshot file: {file_path}")
+    except Exception as e:
+        logger.error(f"Failed to delete memory snapshot file: {file_path} {e}")
+        return
+
+
+# ## 3. Import data
+
+# In[5]:
+
+
+# load in the model and the predictor
+sam2_checkpoint = pathlib.Path("../../../data/models/sam2/sam2_hiera_tiny.pt").resolve()
+model_cfg = "sam2_hiera_t.yaml"
+predictor = build_sam2_video_predictor(model_cfg, sam2_checkpoint)
+
+# set the path to the video
+tiff_dir = pathlib.Path(
+    "../../../2.cellprofiler_ic_processing/illum_directory_test_small/20231017ChromaLive_6hr_4ch_MaxIP/C-02_F0001/"
+).resolve(strict=True)
+video_dir = pathlib.Path(
+    "../../../2.cellprofiler_ic_processing/illum_directory_test_small/20231017ChromaLive_6hr_4ch_MaxIP/C-02_F0001/jpeg/"
+).resolve()
```

Comment:
Just to confirm, are these paths supposed to be to test data? If so, recommend changing to final data in this PR or another PR in the future.

Reply:
Yes, just test data for now. I will change and rerun in a different PR